processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java

```diff
@@ -122,14 +122,13 @@ public Assignment getTargetAssignment(Method mappingMethod, String mappedElement
         ResolvingAttempt attempt = new ResolvingAttempt(
             sourceModel,
             mappingMethod,
-            mappedElement,
             dateFormat,
             numberFormat,
             sourceRHS,
             criteria
         );
 
-        return attempt.getTargetAssignment( sourceRHS.getSourceType(), targetType );
+        return attempt.getTargetAssignment( sourceRHS.getSourceTypeForMatching(), targetType );
```

Comment:
The problem is that the actual `SourceType`  is a collection. The type used for matching is the element. So I want `SourceRHS` to be consistent (for our own understanding, but also whenever we want to make this an templating SPI in the future). In the old solution, the `SourceType` in `SourceRHS` was simply replaced by the type of  the collection element when an ADDER was applicable.

But also here... if you have a better idea, its more than welcome.